`/webdriver/tests/classic/take_screenshot/iframe.py`, `/webdriver/tests/classic/take_element_screenshot/iframe.py` intermittently captures a screenshot of empty page

It's always the reference screenshot, for some reason. The webdriver session loads a page containing an iframe, then blocks on the page being fully loaded, then takes a screenshot of the top level browsing context. The test passes when this screenshot contains real pixels, and fails when the screenshot contains only white pixels.

Another one that occurs often.
```
    FAIL [expected PASS] subtest: test_source_origin[cross_origin]
AssertionError: assert 'iVBORw0KGgoAAAANSUhEUgAAAyAAAAJYCAYAAACadoJwAAArA0lEQVR4Ae3AA6AkWZbG8f937o3IzKdyS2Oubdu2bdu2bdu2bWmMnpZKr54yMyLu+Xa3anqmhztr1a/aNlddddVVV1111VVXXXXVVf/5qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8N/hEuSA1dDZxjrQAAAABJRU5ErkJggg==' == 'iVBORw0KGgoAAAANSUhEUgAAAyAAAAJYCAYAAACadoJwAAAzoElEQVR4Ae3AA6AkWZbG8f937o3IzKdyS2Oubdu2bdu2bdu2bWmMnpZKr54yMyLu+Xa3anqmhztr1a/aNlddddVVV1111VVXXXXVVf/5qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0q/wJ9jrjqfw5/lrnqqquuuuqqq6666qr/pahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKv2QP+Eqe0xsDr8xVV1111VVXXXXVVVddddW/BsG/ZAf4bOClgR74bOCVueqqq6666qqrrrrqqquu+teictVVV1111VVXXXXVVVdd9V+Dyn+GpwC/DdwDdMAjgDcAtrniT4Ff5Io3Be4D/gFYAi8OvD1XPAn4beA+YAY8BngDYAb8KfCLXPHmwFngr4EZ8NrASwG/AvwNMAdeC3hZrrrqqquuuuqqq6666qr/PgT/0Z4E/ADwEOBjgfcDzgPfBay54hWBT+WKPwQeAnwM8KY82xOAHwIeAXwc8B7ArcAPAwZeEfhUrvhz4OHAxwIvDvwc8HPAw4CPBV4c+Hlgl6uuuuqqq6666qqrrrrqvw/Bf7RfBc4ArwdsAKeBtwAuAH/G83oo8GJAD7wi8PZc8WvAGeB1gAVwHfCGwNOBW3lONwCPAGbAKwMGEngUMANeGUjgdq666qqrrrrqqquuuuqq/z4E/5H2gHPAg3lO1wEz4Gk8r2t5XnvAeeDBPKcbueJpPKdreLZNrriGZ9viiiOuuuqqq6666qqrrrrqqv8+VP4jHXHFgue1ARzxvDqe1xFX/CnwpzyvSzynnmcTV/Q8m7jCXHXVVVddddVVV1111VX/faj8R9rgiiXP6wg4wYtmgyteDXgDrrrqqquuuuqqq6666qr/Gwj+I3w/cCuwA5wGbuU53QOsgYfyotkBTgN38by+Cfh7rrrqqquuuuqqq6666qr/fQj+o70hcBb4DeAIOAf8HHASeAVedG8EPAP4feAIOAJ+BUjg0Vx11VVXXXXVVVddddVV//sg2+aF0McJvpJ/2XsDD+aKpwC/BdwLdMAjgDcAtrni74Ef5zl9MjDnOT0V+C3gHmAOPAR4A2AH+Hvgx3m2lwReGvhenu2xwCsC382zPRR4T/7X8meZq6666qqrrrrqqquu+l8K2TYvhD5HXPU/hz/LXHXVVVddddVVV1111f9SVK666qqrrrrqqquuuuqqq/5rULnqqquuuuqqq6666qqrrvqvQeWqq6666qqrrrrqqquuuuq/BpWrrrrqqquuuuqqq6666qr/GlSuuuqqq6666qqrrrrqqqv+axC8KG4DvgQ4y1VXXXXVVVddddVVV1111b8VlReFucL83/WXwM9yxZsDL8+L7teB3+eKdwcezlVXXXXVVVddddVVV131vAheFA8CPgm4hv+7Xhb4dP5tXh/4cK666qqrrrrqqquuuuqqF47KVVddddVVV1111VVXXXXVfw0q/5I/BX6RK94SeFng14Hf54q3BZ4IPBkowEsCbwBUrrgA/ApwOzABDwVeHbgJ+GXgj7niQ4FrgMcBP8oVbwu8JPCnwC9yxZsDZ4G/BmbAawMvBfwK8DfAHHgt4GX5l/0p8AfAEXAj8CY8f08Bfhu4B+iARwBvAGzzL3sK8NvAPUAHPAJ4A2CbK/4U+EWueFPgPuAfgCXw4sDbc9VVV1111VVXXXXVVf9XEPxLXhH4VJ7T6wMfzhW/BbwE8HHAWwJ/Cfwqz/ajQAd8GPCxwHHge7nijYH35Dk9FvhEntMrAp/KFX8OPBz4WODFgZ8Dfg54GPCxwIsDPw/s8sL9HfCLwMsBHwu8EfBrPK8nAT8APAT4WOD9gPPAdwFrXrgnAT8APAT4WOD9gPPAdwFrrnhF4FO54g+BhwAfA7wpV1111VVXXXXVVVdd9X8Nwb/XY4FHAzPgMcDLAn8B7AETcA/wGGATmANvCHT8290APAKYAa8MGEjgUcAMeGUggdt54X4buA54TWABXA+8HM/rV4EzwOsBG8Bp4C2AC8Cf8cL9KnAGeD1gAzgNvAVwAfgzntdDgRcDeuAVgbfnqquuuuqqq6666qqr/i8h+Pe6ged0C9CAu4EK3Aj8OvAPwAgE8An8213Ds21yxTU82xZXHPGCLYHzwC08p5t4TnvAOeDBPKfrgBnwNF6wPeAc8GCe03XADHgaz+tarrrqqquuuuqqq6666v8yKv9eM57TBlfsc8W7A78L/DLwk8BDgdcCbuLfpufZxBU9zyauMC/YPlcseE4LntMRVyx4XhvAES/YEVcseF4bwBHPq+Oqq6666qqrrrrqqqv+LyP491rynI64YpsrFsAbAR8HvA8wAd8FXOAKcUXj2db859rmiiXPacVz2uCKJc/rCNjgBdvgiiXP6wjY4Kqrrrrqqquuuuqqq/6/Ifj3upPndBtQgBuAA+AbebabgLcEGnAXV2xwxQHPdi//uRbAKeA2ntPdPKcd4DRwK8/pHmANPJQXbAc4DdzKc7oHWAMP5aqrrrrqqquuuuqqq/6/Ifj3ejrwBGANPB74S+DlgG2uuA/4I2ANrIA/BypwI1ecAjaBPwOWwDngr/nP99rAPcDvAkvgXuAPeF5vCJwFfgM4As4BPwecBF6BF+4NgbPAbwBHwDng54CTwCtw1VVXXXXVVVddddVV/98g2+aF0JsJfpFnezDw3sA54OuBdwSeCDwBCOAlgDcEKlc8CfgT4G6gAdcCrwM8hGd7GvBLwC5wI/CGwLdyxcOBlwZ+nGd7SeClge/l2R4LvCLw3TzbQ4H35AX7M+D3gUPgOuCNgO/gipuA9+eKpwC/BdwLdMAjgDcAtrni14Hf59leEnhbrngK8FvAvUAHPAJ4A2CbK/4e+HGe0ycDc54vf5a56qqrrrrqqquuuuqq/6WQbfNC6HPE83UO+Hrg3YGHc9V/EX+Wueqqq6666qqrrrrqqv+lqFx11VVXXXXVVVddddVVV/3XoPJv8evA73PF9wMvCbwtV1111VVXXXXVVVddddVVLwyVf4vXB16fq6666qqrrrrqqquuuuqqfw0qV1111VVXXXXVVVddddVV/zUI/iPcBnwJcJarrrrqqquuuuqqq6666qoXhOA/grnCPNvTgM8G7uN/j6cBnw3cx1VXXXXVVVddddVVV131H4/Kf4QHAZ/EVVddddVVV1111VVXXXXVC0Plqquuuuqqq6666qqrrrrqvwaVF8UF4FeA24EJeCjw6sBNwJ8Cv8gVbwm8LPDrwO9zxTdyxXHgo3n+joDfB54A7AGngdcGHs2z/Snwi1zx5sBZ4K+BGfDawEsBvwL8DTAHXgt4WZ7Tk4DfBu4DZsBjgDcAZsCvA7/PFd/IFceBj+aqq6666qqrrrrqqquu+o+BbJsXQp8j+GbgNPAmQAF+G/hL4FO5YgC+EHhL4GW54mnA9wIfClzDC/fLXPGaQAWeAPws8IHANTzbAHwhcB3wesAtwO8AfwS8FPAY4MHA7wJ/BHwkcJwrngD8CPCawCsDl4AfB7aB9wQEPA34XuBDgWv4H8mfZa666qqrrrrqqquuuup/KYJ/yQTcAzwG2ATmwBsCHf9x3hh4Y2AD6IGXBB4G/CnP3w3AI4AZ8MqAgQQeBcyAVwYSuJ1n+zXgDPA6wAK4DnhD4OnArVx11VVXXXXVVVddddVV//kI/iUVuBH4deAfgBEI4BP4z7UAzvP8XcOzbXLFNTzbFlccccUecB54MM/pRq54GlddddVVV1111VVXXXXVfz4qL4p3B34X+GXgJ4GHAq8F3MR/jLPAbwK3A4eAueI6nr+eZxNX9DybuMJcccQVfwr8Kc/rElddddVVV1111VVXXXXVfz4qL4oF8EbAGwF3AL8BfBfwYcBJ/n0a8L3ANvDewEkggJ8G7uE/xgZXvBrwBlx11VVXXXXVVVddddVV/z0I/iUHwDfybDcBbwk04C5eMPGiuQjsAy8OnAaCKyb+4+wAp4G7eF7fBPw9V4irrrrqqquuuuqqq6666j8PwYviPuCPgDWwAv4cqMCNvGBbXHEOOAK+nOfvOLAJ/A1wFpiApwJP4T/WGwHPAH4fOAKOgF8BEng0V2xxxTngCPhy4C7gz4DPBn6dq6666qqrrrrqqquuuurfDtk2L4Q+R/Ak4E+Au4EGXAu8DvAQ4E+BX+TZHgy8N1f8AvB3gIGXBN6M5+9O4FeBu4EF8HBgCTyOKz4euBX4cZ7tJYGXBr6XZ3ss8IrAd/NsDwXekyueCvwWcA8wBx4CvAGww7P9AvB3gIGXBN4M+DPgF4DXBF6X/1b+LHPVVVddddVVV1111VX/SyHb5oXQ54ir/ufwZ5mrrrrqqquuuuqqq676X4rKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf910C2zVVXXXXVVVddddVVV1111X8+KlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DylVXXXXVVVddddVVV1111X8NKlddddVVV1111VVXXXXVVf81qFx11VVXXXXVVVddddVVV/3XoHLVVVddddVVV1111VVXXfVfg8pVV1111VVXXXXVVVddddV/DSpXXXXVVVddddVVV1111VX/NahcddVVV1111VVXXXXVVVf916By1VVXXXXVVVddddVVV131X4PKVVddddVVV1111VVXXXXVfw0qV1111VVXXXXVVVddddVV/zWoXHXVVVddddVVV1111VVX/degctVVV1111VVXXXXVVVdd9V+DfwSjbBDciykQIwAAAABJRU5ErkJggg=='
```

--- a/components/compositing/compositor.rs
+++ b/components/compositing/compositor.rs
@@ -28,11 +28,10 @@ use dpi::PhysicalSize;
 use embedder_traits::{
     CompositorHitTestResult, InputEvent, ScreenshotCaptureError, ShutdownState, ViewportDetails,
 };
-use euclid::{Point2D, Rect, Scale, Size2D, Transform3D};
+use euclid::{Point2D, Scale, Size2D, Transform3D};
 use image::RgbaImage;
 use ipc_channel::ipc::{self, IpcSharedMemory};
 use log::{debug, info, trace, warn};
-use pixels::{CorsStatus, ImageFrame, ImageMetadata, PixelFormat, RasterImage};
 use profile_traits::mem::{
     ProcessReports, ProfilerRegistration, Report, ReportKind, perform_memory_report,
 };
@@ -1184,58 +1183,6 @@ impl IOCompositor {
         self.needs_repaint.set(RepaintReason::empty());
     }
 
-    /// Render the WebRender scene to the shared memory, without updating other state of this
-    /// [`IOCompositor`]. If succesful return the output image in shared memory.
-    pub fn render_to_shared_memory(
-        &mut self,
-        webview_id: WebViewId,
-        page_rect: Option<Rect<f32, CSSPixel>>,
-    ) -> Option<RasterImage> {
-        self.render_inner();
-
-        let size = self.rendering_context.size2d().to_i32();
-        let rect = if let Some(rect) = page_rect {
-            let scale = self
-                .webview_renderers
-                .get(webview_id)
-                .map(WebViewRenderer::device_pixels_per_page_pixel)
-                .unwrap_or_else(|| Scale::new(1.0));
-            let rect = scale.transform_rect(&rect);
-
-            let x = rect.origin.x as i32;
-            // We need to convert to the bottom-left origin coordinate
-            // system used by OpenGL
-            // If dpi > 1, y can be computed to be -1 due to rounding issue, resulting in panic.
-            // https://github.com/servo/servo/issues/39306#issuecomment-3342204869
-            let y = 0.max((size.height as f32 - rect.origin.y - rect.size.height) as i32);
-            let w = rect.size.width as i32;
-            let h = rect.size.height as i32;
-
-            DeviceIntRect::from_origin_and_size(Point2D::new(x, y), Size2D::new(w, h))
-        } else {
-            DeviceIntRect::from_origin_and_size(Point2D::origin(), size)
-        };
-
-        self.rendering_context
-            .read_to_image(rect)
-            .map(|image| RasterImage {
-                metadata: ImageMetadata {
-                    width: image.width(),
-                    height: image.height(),
-                },
-                format: PixelFormat::RGBA8,
-                frames: vec![ImageFrame {
-                    delay: None,
-                    byte_range: 0..image.len(),
-                    width: image.width(),
-                    height: image.height(),
-                }],
-                bytes: ipc::IpcSharedMemory::from_bytes(&image),
-                id: None,
-                cors_status: CorsStatus::Safe,
-            })
-    }
-
     #[servo_tracing::instrument(skip_all)]
     fn render_inner(&mut self) {
         if let Err(err) = self.rendering_context.make_current() {
@@ -1584,6 +1531,16 @@ impl IOCompositor {
         }
     }
 
+    pub fn device_pixels_per_page_pixel(
+        &self,
+        webview_id: WebViewId,
+    ) -> Scale<f32, CSSPixel, DevicePixel> {
+        self.webview_renderers
+            .get(webview_id)
+            .map(WebViewRenderer::device_pixels_per_page_pixel)
+            .unwrap_or_default()
+    }
+
     fn webrender_document(&self) -> DocumentId {
         self.global.borrow().webrender_document
     }
@@ -1644,10 +1601,11 @@ impl IOCompositor {
     pub fn request_screenshot(
         &self,
         webview_id: WebViewId,
+        rect: Option<DeviceRect>,
         callback: Box<dyn FnOnce(Result<RgbaImage, ScreenshotCaptureError>) + 'static>,
     ) {
         self.screenshot_taker
-            .request_screenshot(webview_id, callback);
+            .request_screenshot(webview_id, rect, callback);
         let _ = self.global.borrow().constellation_sender.send(
             EmbedderToConstellationMessage::RequestScreenshotReadiness(webview_id),
         );

--- a/components/compositing/screenshot.rs
+++ b/components/compositing/screenshot.rs
@@ -8,16 +8,19 @@ use std::rc::Rc;
 use base::Epoch;
 use base::id::{PipelineId, WebViewId};
 use embedder_traits::ScreenshotCaptureError;
+use euclid::{Point2D, Size2D};
 use image::RgbaImage;
 use rustc_hash::FxHashMap;
+use webrender_api::units::{DeviceIntRect, DeviceRect};
 
 use crate::IOCompositor;
 use crate::compositor::RepaintReason;
 
 pub(crate) struct ScreenshotRequest {
     webview_id: WebViewId,
-    phase: ScreenshotRequestPhase,
+    rect: Option<DeviceRect>,
     callback: Box<dyn FnOnce(Result<RgbaImage, ScreenshotCaptureError>) + 'static>,
+    phase: ScreenshotRequestPhase,
 }
 
 /// Screenshots requests happen in three phases:
@@ -63,12 +66,14 @@ impl ScreenshotTaker {
     pub(crate) fn request_screenshot(
         &self,
         webview_id: WebViewId,
+        rect: Option<DeviceRect>,
         callback: Box<dyn FnOnce(Result<RgbaImage, ScreenshotCaptureError>) + 'static>,
     ) {
         self.requests.borrow_mut().push(ScreenshotRequest {
             webview_id,
-            phase: ScreenshotRequestPhase::ConstellationRequest,
+            rect,
             callback,
+            phase: ScreenshotRequestPhase::ConstellationRequest,
         });
     }
 
@@ -175,9 +180,25 @@ impl ScreenshotTaker {
                     return None;
                 };
 
+                let viewport_rect = webview_renderer.rect.to_i32();
+                let viewport_size = viewport_rect.size();
+                let rect = screenshot_request.rect.map_or(viewport_rect, |rect| {
+                    // We need to convert to the bottom-left origin coordinate
+                    // system used by OpenGL
+                    // If dpi > 1, y can be computed to be -1 due to rounding issue, resulting in panic.
+                    // https://github.com/servo/servo/issues/39306#issuecomment-3342204869
+                    let x = rect.min.x as i32;
+                    let y = 0.max(
+                        (viewport_size.height as f32 - rect.min.y - rect.size().height) as i32,
+                    );
+                    let w = rect.size().width as i32;
+                    let h = rect.size().height as i32;
+
+                    DeviceIntRect::from_origin_and_size(Point2D::new(x, y), Size2D::new(w, h))
+                });
                 let result = renderer
                     .rendering_context()
-                    .read_to_image(webview_renderer.rect.to_i32())
+                    .read_to_image(rect)
                     .ok_or(ScreenshotCaptureError::CouldNotReadImage);
                 callback(result);
                 None

--- a/components/servo/lib.rs
+++ b/components/servo/lib.rs
@@ -82,13 +82,14 @@ use fonts::SystemFontService;
 use gaol::sandbox::{ChildSandbox, ChildSandboxMethods};
 pub use gleam::gl;
 use gleam::gl::RENDERER;
+pub use image::RgbaImage;
 use ipc_channel::ipc::{self, IpcSender};
 use javascript_evaluator::JavaScriptEvaluator;
 pub use keyboard_types::{
     Code, CompositionEvent, CompositionState, Key, KeyState, Location, Modifiers, NamedKey,
 };
 use layout::LayoutFactoryImpl;
-use log::{Log, Metadata, Record, debug, error, warn};
+use log::{Log, Metadata, Record, debug, warn};
 use media::{GlApi, NativeDisplay, WindowGLContext};
 use net::protocols::ProtocolRegistry;
 use net::resource_thread::new_resource_threads;
@@ -1059,29 +1060,8 @@ impl Servo {
     }
 
     pub fn execute_webdriver_command(&self, command: WebDriverCommandMsg) {
-        if let WebDriverCommandMsg::TakeScreenshot(webview_id, page_rect, response_sender) = command
-        {
-            if let Some(ref rect) = page_rect {
-                if rect.height() == 0.0 || rect.width() == 0.0 {
-                    error!("Taking screenshot of bounding box with zero area");
-                    if let Err(e) = response_sender.send(Err(())) {
-                        error!("Sending reply to create png failed {e:?}");
-                    }
-                    return;
-                }
-            }
-
-            let img = self
-                .compositor
-                .borrow_mut()
-                .render_to_shared_memory(webview_id, page_rect);
-            if let Err(e) = response_sender.send(Ok(img)) {
-                error!("Sending reply to create png failed ({:?}).", e);
-            }
-        } else {
-            self.constellation_proxy
-                .send(EmbedderToConstellationMessage::WebDriverCommand(command));
-        }
+        self.constellation_proxy
+            .send(EmbedderToConstellationMessage::WebDriverCommand(command));
     }
 
     pub fn set_preference(&self, name: &str, value: PrefValue) {

--- a/components/servo/webview.rs
+++ b/components/servo/webview.rs
@@ -20,6 +20,7 @@ use embedder_traits::{
 use euclid::{Point2D, Scale, Size2D};
 use image::RgbaImage;
 use servo_geometry::DeviceIndependentPixel;
+use style_traits::CSSPixel;
 use url::Url;
 use webrender_api::ScrollLocation;
 use webrender_api::units::{DeviceIntPoint, DevicePixel, DeviceRect};
@@ -526,6 +527,13 @@ impl WebView {
             .set_pinch_zoom(self.id(), new_pinch_zoom);
     }
 
+    pub fn device_pixels_per_css_pixel(&self) -> Scale<f32, CSSPixel, DevicePixel> {
+        self.inner()
+            .compositor
+            .borrow()
+            .device_pixels_per_page_pixel(self.id())
+    }
+
     pub fn exit_fullscreen(&self) {
         self.inner()
             .constellation_proxy
@@ -589,9 +597,11 @@ impl WebView {
         );
     }
 
-    /// Asynchronously take a screenshot of the [`WebView`] contents. This method will
-    /// wait until the [`WebView`] is ready before the screenshot is taken. This includes
-    /// waiting for:
+    /// Asynchronously take a screenshot of the [`WebView`] contents, given a `rect` or the whole
+    /// viewport, if no `rect` is given.
+    ///
+    /// This method will wait until the [`WebView`] is ready before the screenshot is taken.
+    /// This includes waiting for:
     ///
     ///  - all frames to fire their `load` event.
     ///  - all render blocking elements, such as stylesheets included via the `<link>`
@@ -606,12 +616,13 @@ impl WebView {
     /// operation.
     pub fn take_screenshot(
         &self,
+        rect: Option<DeviceRect>,
         callback: impl FnOnce(Result<RgbaImage, ScreenshotCaptureError>) + 'static,
     ) {
         self.inner()
             .compositor
             .borrow()
-            .request_screenshot(self.id(), Box::new(callback));
+            .request_screenshot(self.id(), rect, Box::new(callback));
     }
 }
 

--- a/components/shared/embedder/Cargo.toml
+++ b/components/shared/embedder/Cargo.toml
@@ -23,6 +23,7 @@ crossbeam-channel = { workspace = true }
 euclid = { workspace = true }
 http = { workspace = true }
 hyper_serde = { workspace = true }
+image = { workspace = true }
 ipc-channel = { workspace = true }
 keyboard-types = { workspace = true }
 log = { workspace = true }

--- a/components/shared/embedder/webdriver.rs
+++ b/components/shared/embedder/webdriver.rs
@@ -9,12 +9,13 @@ use std::collections::HashMap;
 use base::generic_channel::GenericSender;
 use base::id::{BrowsingContextId, WebViewId};
 use cookie::Cookie;
+use crossbeam_channel::Sender;
 use euclid::default::Rect as UntypedRect;
 use euclid::{Rect, Size2D};
 use hyper_serde::Serde;
+use image::RgbaImage;
 use ipc_channel::ipc::IpcSender;
 use keyboard_types::{CompositionEvent, KeyboardEvent};
-use pixels::RasterImage;
 use rustc_hash::FxHashMap;
 use serde::{Deserialize, Serialize};
 use servo_geometry::DeviceIndependentIntRect;
@@ -23,7 +24,7 @@ use style_traits::CSSPixel;
 use webdriver::error::ErrorStatus;
 use webrender_api::units::DevicePixel;
 
-use crate::{JSValue, MouseButton, MouseButtonAction, TraversalId};
+use crate::{JSValue, MouseButton, MouseButtonAction, ScreenshotCaptureError, TraversalId};
 
 #[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
 pub struct WebDriverMessageId(pub usize);
@@ -73,7 +74,7 @@ impl WebDriverUserPromptAction {
 }
 
 /// Messages to the constellation originating from the WebDriver server.
-#[derive(Debug, Deserialize, Serialize)]
+#[derive(Debug)]
 pub enum WebDriverCommandMsg {
     /// Used in the initialization of the WebDriver server to set the sender for sending responses
     /// back to the WebDriver client. It is set to constellation for now
@@ -144,7 +145,7 @@ pub enum WebDriverCommandMsg {
     TakeScreenshot(
         WebViewId,
         Option<Rect<f32, CSSPixel>>,
-        IpcSender<Result<Option<RasterImage>, ()>>,
+        Sender<Result<RgbaImage, ScreenshotCaptureError>>,
     ),
     /// Create a new webview that loads about:blank. The embedder will use
     /// the provided channels to return the top level browsing context id

--- a/components/webdriver_server/lib.rs
+++ b/components/webdriver_server/lib.rs
@@ -34,12 +34,11 @@ use embedder_traits::{
 };
 use euclid::{Point2D, Rect, Size2D};
 use http::method::Method;
-use image::{DynamicImage, ImageFormat, RgbaImage};
+use image::{DynamicImage, ImageFormat};
 use ipc_channel::ipc::{self, IpcReceiver};
 use keyboard_types::webdriver::{Event as DispatchStringEvent, KeyInputState, send_keys};
 use keyboard_types::{Code, Key, KeyState, KeyboardEvent, Location, NamedKey};
 use log::{debug, error, info};
-use pixels::PixelFormat;
 use serde::de::{Deserializer, MapAccess, Visitor};
 use serde::ser::Serializer;
 use serde::{Deserialize, Serialize};
@@ -2317,61 +2316,33 @@ impl Handler {
             script: "requestAnimationFrame(() => arguments[0]());".to_string(),
             args: None,
         });
-        let webview_id = self.webview_id()?;
-        let mut img = None;
-
-        let interval = 1000;
-        let iterations = 30000 / interval;
-
-        for _ in 0..iterations {
-            let (sender, receiver) = ipc::channel().unwrap();
-            self.send_message_to_embedder(WebDriverCommandMsg::TakeScreenshot(
-                webview_id, rect, sender,
-            ))?;
-
-            match wait_for_ipc_response(receiver)? {
-                Ok(output_img) => {
-                    if let Some(x) = output_img {
-                        img = Some(x);
-                        break;
-                    }
-                },
-                Err(()) => {
-                    return Err(WebDriverError::new(
-                        ErrorStatus::UnknownError,
-                        "The bounding box of element has either 0 width or 0 height",
-                    ));
-                },
-            };
-
-            thread::sleep(Duration::from_millis(interval));
+        if rect.as_ref().is_some_and(Rect::is_empty) {
+            return Err(WebDriverError::new(
+                ErrorStatus::UnknownError,
+                "The requested `rect` has zero width and/or height",
+            ));
         }
+        let webview_id = self.webview_id()?;
+        let (sender, receiver) = crossbeam_channel::unbounded();
+        self.send_message_to_embedder(WebDriverCommandMsg::TakeScreenshot(
+            webview_id, rect, sender,
+        ))?;
 
-        let img = match img {
-            Some(img) => img,
-            None => {
-                return Err(WebDriverError::new(
-                    ErrorStatus::Timeout,
-                    "Taking screenshot timed out",
-                ));
-            },
+        let Ok(result) = receiver.recv() else {
+            return Err(WebDriverError::new(
+                ErrorStatus::UnknownError,
+                "Failed to receive TakeScreenshot response",
+            ));
         };
+        let image = result.map_err(|error| {
+            WebDriverError::new(
+                ErrorStatus::UnknownError,
+                format!("Failed to take screenshot: {error:?}"),
+            )
+        })?;
 
-        // The compositor always sends RGBA pixels.
-        assert_eq!(
-            img.format,
-            PixelFormat::RGBA8,
-            "Unexpected screenshot pixel format"
-        );
-
-        let rgb = RgbaImage::from_raw(
-            img.metadata.width,
-            img.metadata.height,
-            img.first_frame().bytes.to_vec(),
-        )
-        .unwrap();
         let mut png_data = Cursor::new(Vec::new());
-        DynamicImage::ImageRgba8(rgb)
+        DynamicImage::ImageRgba8(image)
             .write_to(&mut png_data, ImageFormat::Png)
             .unwrap();
 

--- a/ports/servoshell/desktop/app.rs
+++ b/ports/servoshell/desktop/app.rs
@@ -24,8 +24,8 @@ use servo::user_content_manager::{UserContentManager, UserScript};
 use servo::webrender_api::ScrollLocation;
 use servo::{
     EmbedderToConstellationMessage, EventLoopWaker, ImeEvent, InputEvent, KeyboardEvent,
-    MouseButtonEvent, MouseMoveEvent, WebDriverCommandMsg, WebDriverScriptCommand,
-    WebDriverUserPromptAction, WheelDelta, WheelEvent, WheelMode,
+    MouseButtonEvent, MouseMoveEvent, ScreenshotCaptureError, WebDriverCommandMsg,
+    WebDriverScriptCommand, WebDriverUserPromptAction, WheelDelta, WheelEvent, WheelMode,
 };
 use url::Url;
 use winit::application::ApplicationHandler;
@@ -613,8 +613,22 @@ impl App {
                 WebDriverCommandMsg::SendAlertText(webview_id, text) => {
                     running_state.set_alert_text_of_newest_dialog(webview_id, text);
                 },
-                WebDriverCommandMsg::TakeScreenshot(..) => {
-                    running_state.servo().execute_webdriver_command(msg);
+                WebDriverCommandMsg::TakeScreenshot(webview_id, rect, result_sender) => {
+                    let Some(webview) = running_state.webview_by_id(webview_id) else {
+                        if let Err(error) =
+                            result_sender.send(Err(ScreenshotCaptureError::WebViewDoesNotExist))
+                        {
+                            warn!("Failed to send response to TakeScreenshot: {error}");
+                        }
+                        continue;
+                    };
+                    let rect =
+                        rect.map(|rect| rect.to_box2d() * webview.device_pixels_per_css_pixel());
+                    webview.take_screenshot(rect, move |result| {
+                        if let Err(error) = result_sender.send(result) {
+                            warn!("Failed to send response to TakeScreenshot: {error}");
+                        }
+                    });
                 },
             };
         }

--- a/ports/servoshell/desktop/app_state.rs
+++ b/ports/servoshell/desktop/app_state.rs
@@ -505,7 +505,7 @@ impl RunningAppState {
             return;
         }
 
-        webview.take_screenshot(move |image| {
+        webview.take_screenshot(None, move |image| {
             achieved_stable_image.set(true);
 
             let Some(output_path) = output_path else {

--- a/ports/servoshell/egl/app_state.rs
+++ b/ports/servoshell/egl/app_state.rs
@@ -973,7 +973,7 @@ impl RunningAppState {
             return;
         }
 
-        webview.take_screenshot(move |image| {
+        webview.take_screenshot(None, move |image| {
             achieved_stable_image.set(true);
 
             let Some(output_path) = output_path else {

--- a/tests/wpt/meta/webdriver/tests/classic/take_element_screenshot/scroll_into_view.py.ini
+++ b/tests/wpt/meta/webdriver/tests/classic/take_element_screenshot/scroll_into_view.py.ini
@@ -1,3 +1,0 @@
-[scroll_into_view.py]
-  [test_scroll_into_view]
-    expected: FAIL

